dask/array/gufunc.py

```diff
@@ -0,0 +1,434 @@
+    core_shapes = merge(output_sizes, *_core_input_shapes)
+
+    loop_input_dimss = [tuple("__loopdim%d__" % d for d in range(max_loopdims - n, max_loopdims)) for n in num_loopdims]
+    input_dimss = [l + c for l, c in zip(loop_input_dimss, core_input_dimss)]
+
+    loop_output_dims = max(loop_input_dimss, key=len) if loop_input_dimss else set()
+
+    ## Assess input args for same size and chunk sizes
+    ### Collect sizes and chunksizes of all dims in all arrays
+    dimsizess = {}
+    chunksizess = {}
+    for dims, shape, chunksizes in zip(input_dimss, input_shapes, input_chunkss):
+        for dim, size, chunksize in zip(dims, shape, chunksizes):
+            _dimsizes = dimsizess.get(dim, [])
+            _dimsizes.append(size)
+            dimsizess[dim] = _dimsizes
+            _chunksizes = chunksizess.get(dim, [])
+            _chunksizes.append(chunksize)
+            chunksizess[dim] = _chunksizes
+    ### Assert correct partitioning, for case:
+    if not allow_rechunk:
+        for dim, sizes in dimsizess.items():
+            ### Check that the arrays have same length for same dimensions or dimension `1`
+            if set(sizes).union({1}) != {1, max(sizes)}:
+                raise ValueError("Dimension `'{}'` with different lengths in arrays".format(dim))
+            chunksizes = chunksizess[dim]
+            ### Check if core dimensions consist of only one chunk
+            if (dim in core_shapes) and (chunksizes[0] < core_shapes[dim]):
+                raise ValueError("Core dimension `'{}'` consists of multiple chunks. To fix, rechunk into a single \
+chunk along this dimension or set `allow_rechunk=True`, but beware that this may increase memory usage \
+significantly.".format(dim))
+            ### Check if loop dimensions consist of same chunksizes, when they have sizes > 1
+            relevant_chunksizes = list(unique(c for s, c in zip(sizes, chunksizes) if s > 1))
+            if len(relevant_chunksizes) > 1:
+                raise ValueError("Dimension `'{}'` with different chunksize present".format(dim))
+
+    ## Apply function - use atop here
+    arginds = list(concat(zip(args, input_dimss)))
+
+    ### Use existing `atop` but only with loopdims to enforce
+    ### concatenation for coredims that appear also at the output
+    ### Modifying `atop` could improve things here.
+    tmp = atop(func, loop_output_dims, *arginds,
+               dtype=int,  # Only dummy dtype, anyone will do
+               concatenate=True,
+               **kwargs)
+
+    ## Prepare output shapes
+    loop_output_shape = tmp.shape
+    loop_output_chunks = tmp.chunks
+    dsk = tmp.__dask_graph__()
+    keys = list(flatten(tmp.__dask_keys__()))
+    _anykey = keys[0]
+    name, token = _anykey[0].split('-')
+
+    ### *) Treat direct output
+    if nout is None:
+        core_output_dimss = [core_output_dimss]
+        output_dtypes = [output_dtypes]
```

Comment:
If you avoided unpacking `core_output_dimss` and `output_dtypes` above, I don't think you would need this step.